lib/mpl_toolkits/mplot3d/art3d.py

```diff
@@ -168,51 +180,70 @@ def line_2d_to_3d(line, zs=0, zdir='z'):
 
 
 def path_to_3d_segment(path, zs=0, zdir='z'):
-    """Convert a path to a 3D segment."""
+    """Convert a path to a 3D segment.
 
+    .. versionchanged :: 3.1
+       Return type changed from a list to a numpy.array
+    """
     zs = np.broadcast_to(zs, len(path))
+
     pathsegs = path.iter_segments(simplify=False, curves=False)
-    seg = [(x, y, z) for (((x, y), code), z) in zip(pathsegs, zs)]
-    seg3d = [juggle_axes(x, y, z, zdir) for (x, y, z) in seg]
-    return seg3d
+    if len(path):
+        xs, ys = zip(*((x, y) for (x, y), code in pathsegs))
+    else:
+        xs, ys = [], []
+
+    seg3d = juggle_axes(xs, ys, zs, zdir)
+    return np.array(seg3d).T
 
 
 def paths_to_3d_segments(paths, zs=0, zdir='z'):
-    """Convert paths from a collection object to 3D segments."""
+    """Convert paths from a collection object to 3D segments.
+
+    .. versionchanged :: 3.1
+       Return type changed from a list to a numpy.array
+    """
 
     zs = np.broadcast_to(zs, len(paths))
+
     segs = [path_to_3d_segment(path, pathz, zdir)
             for path, pathz in zip(paths, zs)]
     return segs
 
 
 def path_to_3d_segment_with_codes(path, zs=0, zdir='z'):
-    """Convert a path to a 3D segment with path codes."""
+    """Convert a path to a 3D segment with path codes.
 
+    .. versionchanged :: 3.1
+       Return type changed from a list to a numpy.array
+    """
     zs = np.broadcast_to(zs, len(path))
-    seg = []
-    codes = []
+
     pathsegs = path.iter_segments(simplify=False, curves=False)
-    for (((x, y), code), z) in zip(pathsegs, zs):
-        seg.append((x, y, z))
-        codes.append(code)
-    seg3d = [juggle_axes(x, y, z, zdir) for (x, y, z) in seg]
-    return seg3d, codes
+    codes = np.empty(len(path))
+    xs, ys, codes = zip(*((x, y, code) for (x, y), code in pathsegs))
```

Comment:
To me it looks like the right thing to do would be to add `path.segments_array`, with a structured dtype of `[('x', float), ('y', float), ('code', int)]` or something. That would save you all these zips, and special cases for when the sequence is empty

Reply:
Absolutely, my intuition is also that `path.iter_segments` might be the bottleneck and that instead of making slice for every vertices group then making a copy with `.flatten()`, using some form of sliding window in numpy (e.g. https://github.com/numpy/numpy/issues/7753) together with reshaping might produce an equivalent result more efficiently (or just returning a view instead of a copy). However, this would require some more in depth refactoring and I feel that this PR already make a significant number of changes with a measurable performance improvements. So I would rather leave additional enhancements to a subsequent PR (along those mentioned in the original PR https://github.com/matplotlib/matplotlib/pull/6085). 

Would that be acceptable for you?

BTW, yes, I need to handle the case of empty sequences, here. Should be fixed now.

Reply:
Also more generally, I have done some profiling when running `plot_trisurf` on a large dataset (to produce the figure in the beginning of this issue), results can be found in [bench_results_trisurf.txt](https://github.com/matplotlib/matplotlib/files/2238704/bench.txt) and I cannot find anything related to `path_to_3d_segment` there. 

The above change aims to prevent calling `juggle_axes` on every single vertex as it was done before. As you suggested there may be a way to prevent memory copies, but a) I still think that it would make sense to optimize `path.iter_segments` before it b) given the profiling results, I'm not sure it's a performance critical part of the code (but I could have missed something)